funflow/src/Control/FunFlow/Steps.hs

```diff
@@ -107,6 +139,46 @@ copyDirToStore = proc (inDir, mbOutDir) -> do
           copyDirRecur inDir p
         ) -< (inDir, outDir)
 
+-- | List the contents of a directory within the store
+listDirContents :: Flow eff ex (CS.Content Dir)
+                               ([CS.Content Dir], [CS.Content File])
+listDirContents = internalManipulateStore
+  ( \store dir -> let
+        item = CS.contentItem dir
+        itemRoot = CS.itemPath store item
+      in do
+        (dirs, files) <- listDir $ CS.contentPath store dir
+        relDirs <- for dirs (stripProperPrefix itemRoot)
+        relFiles <- for files (stripProperPrefix itemRoot)
+        return ( (item :</>) <$> relDirs
+               , (item :</>) <$> relFiles
+               )
+  )
+
+-- | Merge a number of store directories together into a single output directory.
+--   This uses hardlinks to avoid duplicating the data on disk.
+mergeDirs :: Flow eff ex [CS.Content Dir] (CS.Content Dir)
+mergeDirs = proc inDirs -> do
+  paths <- internalManipulateStore
+    ( \store items -> return $ CS.contentPath store <$> items) -< inDirs
+  arr CS.All <<< putInStore
+    ( \d inDirs -> for_ inDirs $ \inDir -> do
+      (subDirs, files) <- listDirRecur inDir
+      for_ subDirs $ \absSubDir -> do
+        relSubDir <- stripProperPrefix inDir absSubDir
+        createDirIfMissing True (d </> relSubDir)
+      for_ files $ \absFile -> do
+        relFile <- stripProperPrefix inDir absFile
+        createLink (toFilePath absFile) (toFilePath $ d </> relFile)
+    ) -< paths
+
+-- | Merge a number of files into a single output directory.
```

Comment:
I think this doesn't behave as advertised and would also copy neighbouring files.
E.g. given items `itemA` and `itemB` with
```
itemA:
- file1
- file2
itemB:
- file1
- file3
```

`mergeFiles -< [ itemA :</> file2, itemB :</> file3]` would produce

```
- file1
- file2
- file3
```

Taking `file1` from `itemB`.

Reply:
Hmm, good point. I'll fix this.